Fixs Posix

gdaow opened the pull request:
That's some fixs / implementations for the Posix platforms (SetThreadName, the CachePlugin code, SharedMemory). I tested under Linux, it should work on Mac modulo the build configuration, as I checked the functions I used were Posix-compliant. I'll try to set up a MacOS virtual machine to fix the build on Mac, and I guess I'll have some more fix soon. I push this already so the work isn't made two times :)

ffulin commented on Code/Core/CoreTest/Tests/TestSharedMemory.cpp:
```diff
@@ -0,0 +1,96 @@
+	bool locked = mutex.TryLock();
+
+	(void)locked;
+	ASSERT(locked);
```
Use TEST_ASSERT. This is not stripped in debug and will remove the need to use (void)x to avoid unused variable wanings.

ffulin commented on Code/Core/CoreTest/Tests/TestSharedMemory.cpp:
```diff
@@ -0,0 +1,96 @@
+void TestSharedMemory::CreateAccessDestroy() const
+{
+
+#if defined(__WINDOWS__)
```
Ideally we'd unify the tests across platforms by having the test compile and run an executable. Probably ok to leave as is and I can re-factor that later.

ffulin commented on Code/Core/Process/SharedMemory.cpp:
```diff
@@ -6,17 +6,67 @@
 
+
+#if defined(__LINUX__) || defined(__APPLE__)
+namespace
```
Note that SharedMemory might not be required for OSX and Linux.
It's currently only used by one feature on Windows, which is to pass info between child processes when FASTBuild is run from Visual Studio. We need two processes because of the way VS terminates compilation.

ffulin commented on Code/Core/Process/SystemMutex.cpp:
```diff
@@ -58,15 +58,16 @@ bool SystemMutex::TryLock()
         if ( rc )
         {
-            if ( errno == EWOULDBLOCK )
+            if ( errno == EAGAIN )
```
When I wrote this code, I verified that the return value was EWOULDBLOCK. Did you get EAGAIN when running a second FASTBuild? Should we be checking for both?

gdaow replied:
You're rigth, I misread some documentation. The good way seems to be checking both :
From http://www.gnu.org/savannah-checkouts/gnu/libc/manual/html_node/Error-Codes.html : 
Portability Note: In many older Unix systems, this condition was indicated by EWOULDBLOCK, which was a distinct error code different from EAGAIN. To make your program portable, you should check for both codes and treat them the same.

ffulin commented on Code/Tools/FBuild/FBuildCore/Cache/CachePlugin.cpp:
```diff
@@ -64,20 +74,18 @@
+        void * func = ::GetProcAddress( (HMODULE)m_DLL, name );
         if ( !func )
         {
             FLOG_WARN( "Missing CachePluginDLL function '%s' (Mangled: %s)", friendlyName, mangledName );
```
This won't compile anymore on Windows - you deleted friendlyName and mangledName

ffulin replied:
This won't compile anymore on Windows, since it references friendlyName and mangledName

ffulin commented on Code/Tools/FBuild/FBuildCore/Cache/CachePlugin.cpp:
```diff
@@ -42,18 +47,23 @@
         }
 
-        m_InitFunc		= (CacheInitFunc)		GetFunction( "CacheInit",		"?CacheInit@@YA_NPEBD@Z" );
-        m_ShutdownFunc	= (CacheShutdownFunc)	GetFunction( "CacheShutdown",	"?CacheShutdown@@YAXXZ"  );
```
These mangled names on Windows ensure we get the correct function signature.

ffulin commented on Code/Tools/FBuild/FBuildCore/Cache/CachePluginInterface.h:
```diff
@@ -4,12 +4,21 @@
 // e.g.
 //#define CACHEPLUGIN_DLL_EXPORT __declspec(dllexport)
 
+extern "C" {
```
This breaks compatibility of the interface for existing cache plugin users.

ffulin commented on Code/fbuild.bff:
```diff
@@ -660,6 +660,7 @@ Alias( 'All+Tests' )
 	.Targets		= { 'All', 'Tests' }
 }
 
+#if __WINDOWS__
```
This shouldn't be necessary. The latest version (v0.87) supports VS project and solution generation on all platforms.

ffulin commented:
Thanks for the pull request - one couple of steps closer to feature parity!

I added some notes to various files.  It's mostly good to go, but there are a couple of issues we'd need to resolve before integrating this.

gdaow commented:
Nice, thanks a lot for the review, I'll get all that fixed. About the mangled names, how should it be handled ? As the mangling is compiler-specific, it seems dangerous to me to hard-code it. (Although I think GCC and Clang mangle the same way).

ffulin commented:
I looked into it a little bit, and it seems switching these functions to extern "C" is probably the only way to go long-term. As you pointed out, the mangling is compiler-specific, and in-fact, it's not even consistent between x86 and x64 (there is already a bug in that the plugin dll only work for x64 even on Windows).

So if we change the functions to be extern C as you proposed, it fixes the problem going forward, but leaves us with the backwards compat issue. I think we can solve that by making the code try the mangled names on x64 Windows only first, and if that fails, fall back to the unmangled names. I'll need to try that to make sure it's going to work, but I think it should be ok.

For your change, perhaps just implement the OS X/Linux part (using the friendlyName and ignoring the mangledName) and leave the Windows as-is. Once it's integrated I'll have a crack at sorting out the Windows bit afterwards.

gdaow commented:
I fixed the things you pointed out. For the mangled name, it seems a good idea to me, I added a default parameter for mangledName, ignored it on Posix, and leaved the Windows code unchanged. Note that backward compatibility will be broken if recompiling plugins, as 'extern "C"' should be added in the plugins code as well. Let me know if there is still things to improve.

ffulin commented on Code/Tools/FBuild/FBuildCore/Cache/CachePlugin.cpp:
```diff
@@ -41,19 +46,29 @@
+            FLOG_WARN( "Cache plugin '%s' load failed (0x%x).", dllName.Get(), dlerror() );
+            return;
+        }
+        dlerror();
```
this dlerror() seems left over from some debugging?

ffulin commented on Code/Tools/FBuild/FBuildCore/Cache/CachePlugin.cpp:
```diff
@@ -64,20 +79,18 @@
     #if defined( __WINDOWS__ )
         ASSERT( m_DLL );
-        void * func = ::GetProcAddress( (HMODULE)m_DLL, mangledName );
+        void * func = ::GetProcAddress( (HMODULE)m_DLL, name );
```
This should read mangledName for now.

poppolopoppo replied:
Build is broken on dev because of this :P
````
Code\Tools\FBuild\FBuildCore\Cache\CachePlugin.cpp(86) : error C2065: 'name' : undeclared identifier
````

ffulin replied:
I'm in the process of fixing it - I'll integrate a new working version soon.

poppolopoppo replied:
Ok thanks

ffulin replied:
Should be fixed now.

ffulin commented:
I've integrated this into the Perforce mainline. These changes will be in the next release (v0.88).

I've submitted a follow-up fix to "dev" which resolved some issues with this change on Windows.

In a future revision, I'll look into breaking the dependency on mangled names in Windows.

ffulin commented:
As a follow up to this, I've intergrated some work-in-progress changes from the Perforce mainline related to completing Linux/OS X support:
 - distributed compilation tests fixed/enabled
 - deleted unused code we don't need to port
 - made the FBuildWorker function as a command line application on OS X/Linux
